oauth2/src/main/java/com/linecorp/armeria/common/auth/oauth2/SerialFuture.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.auth.oauth2;
+
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.util.Exceptions;
+
+/**
+ * Executes submitted {@link Callable} actions asynchronously in sequence.
+ */
+public class SerialFuture {
+
+    private final Queue<Runnable> actions = new LinkedList<>();
+
+    @Nullable
+    private Runnable active;
+
+    @Nullable
+    Executor executor;
+
+    /**
+     * Constructs {@link SerialFuture} with a supplied {@link Executor}.
+     * @param executor An {@link Executor} to execute asynchronous actions.
+     */
+    public SerialFuture(@Nullable Executor executor) {
+        this.executor = executor;
+    }
+
+    /**
+     * Constructs {@link SerialFuture}.
+     */
+    public SerialFuture() {
+        executor = null;
+    }
+
+    /**
+     * Executes submitted {@link Callable} action asynchronously in sequence.
+     * @param action An asynchronous {@link Callable} action to be executed in sequence.
+     *               It returns {@link CompletionStage} that produces the final result when the action
+     *               completes.
+     * @return Returns a new {@link CompletionStage} with the same result or exception as the {@code action},
+     *         that executes the given action when this stage completes.
+     */
+    public synchronized <V> CompletionStage<V> executeAsync(Callable<CompletionStage<V>> action) {
+        final CompletableFuture<V> result = new CompletableFuture<>();
+        actions.add(() -> {
+            final CompletionStage<V> future;
+            try {
+                future = action.call();
+            } catch (Exception e) {
```

Comment:
Catch `Throwable`, since nobody will probably catch when `Error` is thrown here?

Reply:
fixed